File: 백준/Silver/1149. RGB거리/RGB거리.java
import java.util.Scanner;
 
public class Main {
	static final int INF = 987654321;
    static final int m = 3;
    static int n, ret;
    static int[][] a, dp;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();

        a = new int[1004][m];
        dp = new int[1004][m];
        ret = INF;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                a[i][j] = scanner.nextInt();
            }
        }

        for (int i = 0; i < m; i++) {
            ret = Math.min(ret, go(0, i));
        }

        System.out.println(ret);
        scanner.close();
    }

    static int go(int idx, int now) {
        if (idx == n) {
            return 0;
        }

        if (dp[idx][now] != 0) return dp[idx][now];

        dp[idx][now] = INF;

        for (int i = 0; i < m; i++) {
            if (now == i) continue;
            dp[idx][now] = Math.min(dp[idx][now], go(idx + 1, i) + a[idx][i]);
        }

        return dp[idx][now];
    }
}

File: 백준/Silver/1149. RGB거리/RGB거리.cc
#include <bits/stdc++.h>
using namespace std;
const int INF = 987654321, m = 3;
int n, a[1004][m], dp[1004][m], ret = INF;

int go(int idx, int now)
{
    if (idx == n)
    {
        return 0;
    }
    int &ret = dp[idx][now];
    if (ret)
        return ret;
    ret = INF;

    for (int i = 0; i < m; i++)
    {
        if (now == i)
            continue;
        ret = min(ret, go(idx + 1, i) + a[idx][i]);
    }
    return ret;
}
int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n;

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            cin >> a[i][j];
        }
    }

    for (int i = 0; i < m; i++)
    {
        go(0, i);
    }

    for (int i = 0; i < m; i++)
    {
        ret = min(ret, dp[0][i]);
    }

    cout << ret << "\n";
    return 0;
}